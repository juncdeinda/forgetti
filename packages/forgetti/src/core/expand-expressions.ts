import * as t from '@babel/types';
import type * as babel from '@babel/core';
import type { ComponentNode, StateContext } from './types';
import { getHookCallType } from './get-hook-call-type';
import { isPathValid } from './checks';
import unwrapNode from './unwrap-node';

function isStatementValid(path: babel.NodePath): boolean {
  if (path) {
    switch (path.node.type) {
      case 'ForInStatement':
      case 'ForOfStatement':
      case 'ForStatement':
      case 'WhileStatement':
      case 'DoWhileStatement':
        return false;
      default:
        return true;
    }
  }
  return false;
}

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      isPathValid(current, t.isConditionalExpression) &&
      (current.get('consequent').node === prev.node ||
        current.get('alternate').node === prev.node)
    ) {
      return false;
    }
    if (
      isPathValid(current, t.isLogicalExpression) &&
      current.get('right').node === prev.node
    ) {
      return false;
    }
    prev = current;
    current = current.parentPath;
  }
  return true;
}

const UNDEFINED_LITERAL = t.unaryExpression('void', t.numericLiteral(0));

function transformOptionalCall(
  path: babel.NodePath<t.OptionalCallExpression>,
): t.Expression {
  const unwrappedID = unwrapNode(path.node.callee, t.isIdentifier);
  if (unwrappedID) {
    return t.conditionalExpression(
      t.binaryExpression('==', unwrappedID, t.nullLiteral()),
      UNDEFINED_LITERAL,
      t.callExpression(unwrappedID, path.node.arguments),
    );
  }
  const temp = path.scope.generateUidIdentifier('nullish');
  path.scope.push({
    kind: 'let',
    id: temp,
  });
  const unwrappedCallee =
    unwrapNode(path.node.callee, t.isMemberExpression) ||
    unwrapNode(path.node.callee, t.isOptionalMemberExpression);
  if (unwrappedCallee) {
    let unwrappedObject = unwrapNode(unwrappedCallee.object, t.isIdentifier);
    if (!unwrappedObject) {
      unwrappedObject = path.scope.generateUidIdentifier('object');
      path.scope.push({
        kind: 'let',
        id: unwrappedObject,
      });
      unwrappedCallee.object = t.assignmentExpression(
        '=',
        unwrappedObject,
        unwrappedCallee.object,
      );
    }
    return t.conditionalExpression(
      t.binaryExpression(
        '==',
        t.assignmentExpression('=', temp, unwrappedCallee),
        t.nullLiteral(),
      ),
      UNDEFINED_LITERAL,
      t.callExpression(t.memberExpression(temp, t.identifier('call')), [
        unwrappedObject,
        ...path.node.arguments,
      ]),
    );
  }
  return t.conditionalExpression(
    t.binaryExpression(
      '==',
      t.assignmentExpression('=', temp, path.node.callee),
      t.nullLiteral(),
    ),
    UNDEFINED_LITERAL,
    t.callExpression(temp, path.node.arguments),
  );
}

function transformOptionalMember(
  path: babel.NodePath<t.OptionalMemberExpression>,
): t.Expression {
  const unwrappedID = unwrapNode(path.node.object, t.isIdentifier);
  if (unwrappedID) {
    return t.conditionalExpression(
      t.binaryExpression('==', unwrappedID, t.nullLiteral()),
      UNDEFINED_LITERAL,
      t.memberExpression(unwrappedID, path.node.property, path.node.computed),
    );
  }
  const temp = path.scope.generateUidIdentifier('nullish');
  path.scope.push({
    kind: 'let',
    id: temp,
  });
  return t.conditionalExpression(
    t.binaryExpression(
      '==',
      t.assignmentExpression('=', temp, path.node.object),
      t.nullLiteral(),
    ),
    UNDEFINED_LITERAL,
    t.memberExpression(temp, path.node.property, path.node.computed),
  );
}

export function expandExpressions(
  ctx: StateContext,
  path: babel.NodePath<ComponentNode>,
): void {
  if (
    path.node.type === 'ArrowFunctionExpression' &&
    path.node.body.type !== 'BlockStatement'
  ) {
    path.node.body = t.blockStatement([t.returnStatement(path.node.body)]);
  }
  path.traverse({
    OptionalCallExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (parent === path && statement) {
        p.replaceWith(transformOptionalCall(p));
      }
    },
    OptionalMemberExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (parent === path && statement) {
        p.replaceWith(transformOptionalMember(p));
      }
    },
    AssignmentExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (
        parent === path &&
        statement &&
        isStatementValid(statement) &&
        isInValidExpression(p) &&
        !isPathValid(p.parentPath, t.isStatement) &&
        !isPathValid(p.parentPath, t.isVariableDeclarator)
      ) {
        const id = p.scope.generateUidIdentifier('hoisted');
        statement.insertBefore(
          t.variableDeclaration('let', [t.variableDeclarator(id, p.node)]),
        );
        p.replaceWith(id);
      }
    },
    CallExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (
        parent === path &&
        statement &&
        !isPathValid(p.parentPath, t.isStatement) &&
        !isPathValid(p.parentPath, t.isVariableDeclarator)
      ) {
        const hookType = getHookCallType(ctx, p);
        if (hookType === 'custom' || hookType === 'effect') {
          const id = p.scope.generateUidIdentifier('hoisted');
          statement.insertBefore(
            t.variableDeclaration('let', [t.variableDeclarator(id, p.node)]),
          );
          p.replaceWith(id);
        }
      }
    },
  });

  path.scope.crawl();
}
