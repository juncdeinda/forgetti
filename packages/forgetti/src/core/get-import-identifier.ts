import type * as babel from '@babel/core';
import type * as t from '@babel/types';
import { addDefault, addNamed } from '@babel/helper-module-imports';
import type { ImportDefinition } from './presets';
import type { StateContext } from './types';

export default function getImportIdentifier(
  ctx: StateContext,
  path: babel.NodePath,
  definition: ImportDefinition,
): t.Identifier {
  const name = definition.kind === 'default' ? 'default' : definition.name;
  const target = `${definition.source}[${name}]`;
  const current = ctx.imports.get(target);
  if (current) {
    return current;
  }
  const newID =
    definition.kind === 'named'
      ? addNamed(path, definition.name, definition.source)
      : addDefault(path, definition.source);
  ctx.imports.set(target, newID);
  return newID;
}
