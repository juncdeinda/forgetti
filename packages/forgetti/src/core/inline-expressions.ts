import type * as babel from '@babel/core';
import * as t from '@babel/types';
import type { ComponentNode } from './types';
import { isPathValid } from './checks';

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      isPathValid(current, t.isConditionalExpression) &&
      (current.get('consequent').node === prev.node ||
        current.get('alternate').node === prev.node)
    ) {
      return false;
    }
    if (
      isPathValid(current, t.isLogicalExpression) &&
      current.get('right').node === prev.node
    ) {
      return false;
    }
    prev = current;
    current = current.parentPath;
  }
  return true;
}

function inlineExpression(
  parentPath: babel.NodePath<ComponentNode>,
  path: babel.NodePath<t.Expression>,
) {
  if (
    path.getFunctionParent() === parentPath &&
    isPathValid(path, t.isIdentifier)
  ) {
    const binding = path.scope.getBinding(path.node.name);
    if (binding?.referenced && binding.referencePaths.length === 1) {
      switch (binding.kind) {
        case 'const':
        case 'let':
        case 'var': {
          // move the node to the reference
          const ref = binding.referencePaths[0];
          if (
            isInValidExpression(ref) &&
            isPathValid(binding.path, t.isVariableDeclarator) &&
            binding.path.node.init &&
            isPathValid(binding.path.get('id'), t.isIdentifier) &&
            binding.path.scope.getBlockParent() === ref.scope.getBlockParent()
          ) {
            ref.replaceWith(binding.path.node.init);
            binding.path.remove();
          }
          break;
        }
      }
    }
  }
}

export function inlineExpressions(path: babel.NodePath<ComponentNode>): void {
  path.traverse({
    Expression(p) {
      inlineExpression(path, p);
    },
  });
  path.scope.crawl();
}
