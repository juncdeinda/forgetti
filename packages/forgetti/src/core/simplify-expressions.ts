import type * as babel from '@babel/core';
import * as t from '@babel/types';
import type { ComponentNode } from './types';

type LiteralState = 'truthy' | 'falsy' | 'nullish' | 'indeterminate';

function getBooleanishState(node: t.Expression): LiteralState {
  switch (node.type) {
    case 'BooleanLiteral':
      return node.value ? 'truthy' : 'falsy';
    case 'NumericLiteral':
      return node.value === 0 ? 'falsy' : 'truthy';
    case 'NullLiteral':
      return 'nullish';
    case 'StringLiteral':
      return node.value === '' ? 'falsy' : 'truthy';
    case 'BigIntLiteral':
      return node.value === '0' ? 'falsy' : 'truthy';
    case 'Identifier': {
      return 'indeterminate';
    }
    // case 'ArrayExpression':
    // case 'FunctionExpression':
    // case 'ArrowFunctionExpression':
    // case 'RegExpLiteral':
    // case 'ObjectExpression':
    // case 'RecordExpression':
    // case 'TupleExpression':
    //   return 'truthy';
    case 'ParenthesizedExpression':
    case 'TypeCastExpression':
    case 'TSAsExpression':
    case 'TSSatisfiesExpression':
    case 'TSNonNullExpression':
    case 'TSTypeAssertion':
    case 'TSInstantiationExpression':
      return getBooleanishState(node.expression);
    default:
      return 'indeterminate';
  }
}

export function simplifyExpressions(path: babel.NodePath<ComponentNode>): void {
  path.traverse({
    ConditionalExpression: {
      exit(p) {
        const state = getBooleanishState(p.node.test);
        if (state === 'truthy') {
          p.replaceWith(p.node.consequent);
        } else if (state !== 'indeterminate') {
          p.replaceWith(p.node.alternate);
        }
      },
    },
    LogicalExpression: {
      exit(p) {
        switch (getBooleanishState(p.node.left)) {
          case 'nullish': {
            p.replaceWith(
              p.node.operator === '??' ? p.node.right : p.node.left,
            );
            break;
          }
          case 'falsy': {
            p.replaceWith(
              p.node.operator === '||' ? p.node.right : p.node.left,
            );
            break;
          }
          case 'truthy': {
            p.replaceWith(
              p.node.operator === '&&' ? p.node.right : p.node.left,
            );
            break;
          }
        }
      },
    },
    UnaryExpression: {
      exit(p) {
        const state = getBooleanishState(p.node.argument);
        switch (p.node.operator) {
          case 'void': {
            if (state !== 'indeterminate') {
              p.replaceWith(t.unaryExpression('void', t.numericLiteral(0)));
            }
            break;
          }
          case '!': {
            if (state === 'truthy') {
              p.replaceWith(t.booleanLiteral(false));
            } else if (state !== 'indeterminate') {
              p.replaceWith(t.booleanLiteral(true));
            }
            break;
          }
        }
      },
    },
    IfStatement: {
      exit(p) {
        const state = getBooleanishState(p.node.test);
        if (state === 'truthy') {
          p.replaceWith(p.node.consequent);
        } else if (state === 'indeterminate') {
          // TODO Should simplify IfStatement?
        } else if (p.node.alternate) {
          p.replaceWith(p.node.alternate);
        } else {
          p.remove();
        }
      },
    },
    WhileStatement: {
      exit(p) {
        if (getBooleanishState(p.node.test) === 'falsy') {
          p.remove();
        }
      },
    },
  });
}
