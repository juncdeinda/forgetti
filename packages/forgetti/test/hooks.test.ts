import * as babel from '@babel/core';
import { describe, it } from 'vitest';
import type { Options } from '../src';
import plugin from '../src';

const options: Options = {
  preset: 'react',
};

async function compile(code: string): Promise<string> {
  const result = await babel.transformAsync(code, {
    plugins: [[plugin, options]],
    parserOpts: {
      plugins: ['jsx'],
    },
  });

  return result?.code ?? '';
}

describe('hooks', () => {
  it('should optimize useRef', async ({ expect }) => {
    const code = `
import { useRef } from 'react';
function Example(props) {
  return useRef(props.value);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useMemo', async ({ expect }) => {
    const code = `
import { useMemo } from 'react';
function Example(props) {
  return useMemo(() => props.value(), [props.value]);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useRef and useMemo', async ({ expect }) => {
    const code = `
import { useRef, useMemo } from 'react';
function Example(props) {
  const aRef = useRef(props.foo);
  return useMemo(() => props.bar(), [props.bar]);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useMemo with 0 dependencies', async ({ expect }) => {
    const code = `
import { useMemo } from 'react';
function Example(props) {
  return useMemo(() => props.value(), []);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useMemo with auto dependencies', async ({ expect }) => {
    const code = `
import { useMemo } from 'react';
function Example(props) {
  return useMemo(() => props.value());
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useCallback', async ({ expect }) => {
    const code = `
import { useCallback } from 'react';
function Example(props) {
  return useCallback(() => props.value(), [props.value]);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useCallback with 0 dependencies', async ({ expect }) => {
    const code = `
import { useCallback } from 'react';
function Example(props) {
  return useCallback(() => props.value(), []);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useCallback with auto dependencies', async ({
    expect,
  }) => {
    const code = `
import { useCallback } from 'react';
function Example(props) {
  return useCallback(() => props.value());
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useEffect', async ({ expect }) => {
    const code = `
import { useEffect } from 'react';
function Example(props) {
  useEffect(() => props.value(), [props.example]);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useEffect with 0 dependencies', async ({ expect }) => {
    const code = `
import { useEffect } from 'react';
function Example(props) {
  useEffect(() => props.value(), []);
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize useEffect with auto dependencies', async ({ expect }) => {
    const code = `
import { useEffect } from 'react';
function Example(props) {
  useEffect(() => props.value());
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should correct transform nested hooks call (issue #14)', async ({
    expect,
  }) => {
    const code = `
import { useDeferredValue } from 'react';
import { useAtomValue } from 'jotai';
import { stateAtom } from 'whatever';
function Example(props) {
  return useDeferredValue(useAtomValue(stateAtom));
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should correct transform any nested hooks call', async ({ expect }) => {
    const code = `
import { useA, useB, useC, useD, useE, useF, useG, useH } from 'whatever'

function Example(props) {
  let a = null;
  return useA(
    useB(),
    [useC(), 'array'],
    { d: useD(), [useE()]: useF(), ...useG() },
    \`testA\${useH()}testB\`,
    useI() === useJ(),
    (a = useK()),
    ...useJ()
  );
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should correct transform derived hooks call', async ({ expect }) => {
    const code = `
import { useA, useB, useC } from 'whatever'

function Example(props) {
  let a = null;
  return {
    [useA()]: useB(),
    ...useC(),
    [\`testA\${useH()}testB\`]: useI() === useJ(),
    a: useK() ? 'a' : 'b',
    b: <div>{useL()}</div>,
    c: <>{useM()}{useN()}</>
  }
}
`;
    expect(await compile(code)).toMatchSnapshot();
  });
});
