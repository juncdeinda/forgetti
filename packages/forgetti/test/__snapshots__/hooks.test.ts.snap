// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hooks > should correct transform any nested hooks call 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { useA, useB, useC, useD, useE, useF, useG, useH } from 'whatever';
function Example(props) {
  let _cache = _$$cache(_useMemo, 20),
    a = null,
    _hoisted = useB(),
    _hoisted2 = useC(),
    _hoisted3 = useD(),
    _hoisted4 = useE(),
    _hoisted5 = useF(),
    _hoisted6 = useG(),
    _hoisted7 = useH(),
    _hoisted8 = useI(),
    _hoisted9 = useJ(),
    _hoisted12 = useK(),
    _hoisted10 = a = _$$equals(_cache, 0, _hoisted12) ? _cache[0] : _cache[0] = _hoisted12,
    _hoisted11 = useJ(),
    _equals3 = _$$equals(_cache, 2, _hoisted2),
    _value3 = _equals3 ? _cache[2] : _cache[2] = _hoisted2,
    _value4 = _equals3 ? _cache[3] : _cache[3] = [_value3, 'array'],
    _equals5 = _$$equals(_cache, 5, _hoisted3),
    _value6 = _equals5 ? _cache[5] : _cache[5] = _hoisted3,
    _equals6 = _$$equals(_cache, 6, _hoisted4),
    _value7 = _equals6 ? _cache[6] : _cache[6] = _hoisted4,
    _equals7 = _$$equals(_cache, 7, _hoisted5),
    _value8 = _equals7 ? _cache[7] : _cache[7] = _hoisted5,
    _equals8 = _$$equals(_cache, 8, _hoisted6),
    _value9 = _equals8 ? _cache[8] : _cache[8] = _hoisted6,
    _value10 = _equals5 && _equals6 && _equals7 && _equals8 ? _cache[9] : _cache[9] = {
      d: _value6,
      [_value7]: _value8,
      ..._value9
    },
    _equals11 = _$$equals(_cache, 11, _hoisted7),
    _value12 = _equals11 ? _cache[11] : _cache[11] = _hoisted7,
    _value13 = _equals11 ? _cache[12] : _cache[12] = \`testA\${_value12}testB\`,
    _equals13 = _$$equals(_cache, 14, _hoisted8),
    _value15 = _equals13 ? _cache[14] : _cache[14] = _hoisted8,
    _equals14 = _$$equals(_cache, 15, _hoisted9),
    _value16 = _equals14 ? _cache[15] : _cache[15] = _hoisted9,
    _value17 = _equals13 && _equals14 ? _cache[16] : _cache[16] = _value15 === _value16;
  return useA(_$$equals(_cache, 1, _hoisted) ? _cache[1] : _cache[1] = _hoisted, _$$equals(_cache, 4, _value4) ? _cache[4] : _cache[4] = _value4, _$$equals(_cache, 10, _value10) ? _cache[10] : _cache[10] = _value10, _$$equals(_cache, 13, _value13) ? _cache[13] : _cache[13] = _value13, _$$equals(_cache, 17, _value17) ? _cache[17] : _cache[17] = _value17, _$$equals(_cache, 18, _hoisted10) ? _cache[18] : _cache[18] = _hoisted10, ...(_$$equals(_cache, 19, _hoisted11) ? _cache[19] : _cache[19] = _hoisted11));
}"
`;

exports[`hooks > should correct transform derived hooks call 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { memo as _memo } from \\"react\\";
import { $$memo as _$$memo } from \\"forgetti/runtime\\";
const _Example = _$$memo(_memo, \\"_Example\\", _values => <div>{_values[0]}</div>),
  _Example2 = _$$memo(_memo, \\"_Example2\\", _values2 => <>{_values2[0]}{_values2[1]}</>);
import { useA, useB, useC } from 'whatever';
function Example(props) {
  let _cache = _$$cache(_useMemo, 24),
    a = null,
    _hoisted = useA(),
    _hoisted2 = useB(),
    _hoisted3 = useC(),
    _hoisted4 = useH(),
    _hoisted5 = useI(),
    _hoisted6 = useJ(),
    _hoisted7 = useK(),
    _hoisted8 = useL(),
    _hoisted9 = useM(),
    _hoisted10 = useN(),
    _equals = _$$equals(_cache, 0, _hoisted),
    _value = _equals ? _cache[0] : _cache[0] = _hoisted,
    _equals2 = _$$equals(_cache, 1, _hoisted2),
    _value2 = _equals2 ? _cache[1] : _cache[1] = _hoisted2,
    _equals3 = _$$equals(_cache, 2, _hoisted3),
    _value3 = _equals3 ? _cache[2] : _cache[2] = _hoisted3,
    _equals4 = _$$equals(_cache, 3, _hoisted4),
    _value4 = _equals4 ? _cache[3] : _cache[3] = _hoisted4,
    _value5 = _equals4 ? _cache[4] : _cache[4] = \`testA\${_value4}testB\`,
    _equals5 = _$$equals(_cache, 5, _value5),
    _value6 = _equals5 ? _cache[5] : _cache[5] = _value5,
    _equals6 = _$$equals(_cache, 6, _hoisted5),
    _value7 = _equals6 ? _cache[6] : _cache[6] = _hoisted5,
    _equals7 = _$$equals(_cache, 7, _hoisted6),
    _value8 = _equals7 ? _cache[7] : _cache[7] = _hoisted6,
    _value9 = _equals6 && _equals7 ? _cache[8] : _cache[8] = _value7 === _value8,
    _equals9 = _$$equals(_cache, 9, _value9),
    _value10 = _equals9 ? _cache[9] : _cache[9] = _value9,
    _value11;
  if (_$$equals(_cache, 10, _hoisted7) ? _cache[10] : _cache[10] = _hoisted7) {
    _value11 = 'a';
  } else {
    _value11 = 'b';
  }
  let _equals11 = _$$equals(_cache, 11, _value11),
    _value13 = _equals11 ? _cache[11] : _cache[11] = _value11,
    _equals12 = _$$equals(_cache, 12, _hoisted8),
    _value14 = _equals12 ? _cache[12] : _cache[12] = _hoisted8,
    _value15 = _equals12 ? _cache[13] : _cache[13] = [_value14],
    _equals13 = _$$equals(_cache, 14, _value15),
    _value16 = _equals13 ? _cache[14] : _cache[14] = _value15,
    _value17 = _equals13 ? _cache[15] : _cache[15] = /*@forgetti jsx*/<_Example v={_value16} />,
    _equals14 = _$$equals(_cache, 16, _value17),
    _value18 = _equals14 ? _cache[16] : _cache[16] = _value17,
    _equals15 = _$$equals(_cache, 17, _hoisted9),
    _value19 = _equals15 ? _cache[17] : _cache[17] = _hoisted9,
    _equals16 = _$$equals(_cache, 18, _hoisted10),
    _value20 = _equals16 ? _cache[18] : _cache[18] = _hoisted10,
    _value21 = _equals15 && _equals16 ? _cache[19] : _cache[19] = [_value19, _value20],
    _equals18 = _$$equals(_cache, 20, _value21),
    _value22 = _equals18 ? _cache[20] : _cache[20] = _value21,
    _value23 = _equals18 ? _cache[21] : _cache[21] = /*@forgetti jsx*/<_Example2 v={_value22} />,
    _equals19 = _$$equals(_cache, 22, _value23),
    _value24 = _equals19 ? _cache[22] : _cache[22] = _value23;
  return _equals && _equals2 && _equals3 && _equals5 && _equals9 && _equals11 && _equals14 && _equals19 ? _cache[23] : _cache[23] = {
    [_value]: _value2,
    ..._value3,
    [_value6]: _value10,
    a: _value13,
    b: _value18,
    c: _value24
  };
}"
`;

exports[`hooks > should correct transform nested hooks call (issue #14) 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { useDeferredValue } from 'react';
import { useAtomValue } from 'jotai';
import { stateAtom } from 'whatever';
function Example(props) {
  let _cache = _$$cache(_useMemo, 2),
    _hoisted = useAtomValue(0 in _cache ? _cache[0] : _cache[0] = stateAtom);
  return useDeferredValue(_$$equals(_cache, 1, _hoisted) ? _cache[1] : _cache[1] = _hoisted);
}"
`;

exports[`hooks > should optimize useCallback 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { useCallback } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 5),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.value,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals2 ? _cache[3] : _cache[3] = [_value3];
  return _equals2 ? _cache[4] : _cache[4] = () => props.value();
}"
`;

exports[`hooks > should optimize useCallback with 0 dependencies 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { useCallback } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 2),
    _value = 0 in _cache ? _cache[0] : _cache[0] = [];
  return 1 in _cache ? _cache[1] : _cache[1] = () => props.value();
}"
`;

exports[`hooks > should optimize useCallback with auto dependencies 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { useCallback } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 2),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props;
  return _equals ? _cache[1] : _cache[1] = () => props.value();
}"
`;

exports[`hooks > should optimize useEffect 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { useEffect } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 4),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.example,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2;
  useEffect(() => props.value(), [_equals2 ? _cache[3] : _cache[3] = [_value3]]);
}"
`;

exports[`hooks > should optimize useEffect with 0 dependencies 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { useEffect } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 1);
  useEffect(() => props.value(), [0 in _cache ? _cache[0] : _cache[0] = []]);
}"
`;

exports[`hooks > should optimize useEffect with auto dependencies 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { useEffect } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 2),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = () => props.value();
  useEffect(_value2, [_value2]);
}"
`;

exports[`hooks > should optimize useMemo 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { useMemo } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 5),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = _value.value,
    _equals2 = _$$equals(_cache, 2, _value2),
    _value3 = _equals2 ? _cache[2] : _cache[2] = _value2,
    _value4 = _equals2 ? _cache[3] : _cache[3] = [_value3];
  return _equals2 ? _cache[4] : _cache[4] = (() => props.value())();
}"
`;

exports[`hooks > should optimize useMemo with 0 dependencies 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { useMemo } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 2),
    _value = 0 in _cache ? _cache[0] : _cache[0] = [];
  return 1 in _cache ? _cache[1] : _cache[1] = (() => props.value())();
}"
`;

exports[`hooks > should optimize useMemo with auto dependencies 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { useMemo } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 3),
    _equals = _$$equals(_cache, 0, props),
    _value = _equals ? _cache[0] : _cache[0] = props,
    _value2 = _equals ? _cache[1] : _cache[1] = () => props.value();
  return _equals ? _cache[2] : _cache[2] = _value2();
}"
`;

exports[`hooks > should optimize useRef 1`] = `
"import { useRef as _useRef } from \\"react\\";
import { $$ref as _$$ref } from \\"forgetti/runtime\\";
import { useRef } from 'react';
function Example(props) {
  let _ref = _$$ref(_useRef, 1);
  return 0 in _ref ? _ref[0] : _ref[0] = {
    current: props.value
  };
}"
`;

exports[`hooks > should optimize useRef and useMemo 1`] = `
"import { useRef as _useRef } from \\"react\\";
import { $$ref as _$$ref } from \\"forgetti/runtime\\";
import { useMemo as _useMemo } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
import { useRef, useMemo } from 'react';
function Example(props) {
  let _cache = _$$cache(_useMemo, 5),
    _ref = _$$ref(_useRef, 1);
  const aRef = 0 in _ref ? _ref[0] : _ref[0] = {
    current: props.foo
  };
  let _equals = _$$equals(_cache, 0, props),
    _value2 = _equals ? _cache[0] : _cache[0] = props,
    _value3 = _equals ? _cache[1] : _cache[1] = _value2.bar,
    _equals2 = _$$equals(_cache, 2, _value3),
    _value4 = _equals2 ? _cache[2] : _cache[2] = _value3,
    _value5 = _equals2 ? _cache[3] : _cache[3] = [_value4];
  return _equals2 ? _cache[4] : _cache[4] = (() => props.bar())();
}"
`;
